### ISSUE TITLE
How to make diagrams stable

### ISSUE THREAD

**Issue by ray440** [mermaid-js/mermaid#2833]
I'm doing pixel-based unit tests of mermaid diagrams.  The tests indicate that the diagrams can be (drastically) different on each run.  
How can I stabilize the layout & render of flowchart diagrams so they are identical on each run?

**Comment by ray440** [comment]
Here's an example of the problem. 
![diff3](https://user-images.githubusercontent.com/10067289/158892721-4d30b830-32f1-4982-99d8-bdb7d19b1f33.gif)

And a very subtle example:
![diff4](https://user-images.githubusercontent.com/10067289/158892804-234e2aff-25ad-45da-9a8f-59006d76f7bc.gif)

In both cases the images should be the same.

**Comment by sidharthv96** [comment]
This is a problem for us with our UI tests too. 
Sadly, the rendering algorithms use external libraries, so it might be a bit difficult to narrow down the problem. 
Any help in this area is appreciated.

**Comment by ray440** [comment]
> This is a problem for us with our UI tests too. Sadly, the rendering algorithms use external libraries, so it might be a bit difficult to narrow down the problem. 

Ah, I see. Perhaps we can contact the external parties?

> Any help in this area is appreciated.

What kinda help do you need?

**Comment by sidharthv96** [comment]
> What kinda help do you need?

Firstly, we need to properly analyse why this is happening. 
Once we know if the issue is in mermaid or in the libraries, maybe make a small POC that replicates the issue without the bulk of mermaid code. 
Then explore what kind of fixes are possible.



PS: Graph layout is a very hard problem.

**Comment by jgreywolf** [comment]
@ray440 @sidharthv96  Is this still an issue with most recent version of Mermaid?  I figured it would be good to ask since it has been almost three years :)

**Comment by ray440** [comment]
> Is this still an issue with most recent version of Mermaid?

Yeah, it's been a while. I suspect the problem still exists because mermaid depends on external libs.
